cpp/src/c_api/cugraph_api.cpp

```diff
@@ -0,0 +1,360 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cugraph_c/cugraph_api.h>
+#include <cugraph/api_helpers.hpp>
+
+#include <cugraph/visitors/rw_visitor.hpp>
+
+#include <cugraph/visitors/erased_api.hpp>
+
+#include <raft/handle.hpp>
+
+#include <rmm/device_buffer.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <iostream>
+
+namespace helpers {
+void* raw_device_ptr(cugraph_device_buffer_t* ptr_buf)
+{
+  if (!ptr_buf || !ptr_buf->data_)
+    return nullptr;
+  else {
+    rmm::device_buffer* ptr_d_buffer = reinterpret_cast<rmm::device_buffer*>(ptr_buf->data_);
+    return ptr_d_buffer->data();
+  }
+}
+
+cugraph::visitors::graph_envelope_t const& extract_graph_envelope(
+  cugraph_graph_envelope_t* ptr_graph)
+{
+  using namespace cugraph::visitors;
+
+  return_t::base_return_t* p_base_ret = reinterpret_cast<return_t::base_return_t*>(ptr_graph);
+  return_t::generic_return_t<graph_envelope_t>* p_typed_ret =
+    dynamic_cast<return_t::generic_return_t<graph_envelope_t>*>(p_base_ret);
+
+  graph_envelope_t const& graph_envelope = p_typed_ret->get();
+
+  return graph_envelope;
+}
+
+template <size_t index>
+cugraph_error_t extract_rw_result(cugraph_rw_ret_t* p_rw_ret, cugraph_device_buffer_t* p_d_buf)
+{
+  if (!p_rw_ret)
+    return CUGRAPH_ALLOC_ERROR;
+  else
+    try {
+      using namespace cugraph::visitors;
+      using actual_ret_t = std::tuple<rmm::device_buffer, rmm::device_buffer, rmm::device_buffer>;
+      return_t* p_ret    = reinterpret_cast<return_t*>(p_rw_ret->p_erased_ret);
+
+      auto const& tpl_erased = p_ret->get<actual_ret_t>();
+
+      rmm::device_buffer const& d_vertex = std::get<index>(tpl_erased);
+      p_d_buf->data_                     = const_cast<void*>(static_cast<void const*>(&d_vertex));
+      p_d_buf->size_                     = d_vertex.size();
+
+      return CUGRAPH_SUCCESS;
+    } catch (...) {
+      return CUGRAPH_UNKNOWN_ERROR;
+    }
+}
+}  // namespace helpers
+
+extern "C" {
+int data_type_sz[] = {4, 8, 4, 8};
+}
+
+bool_t runtime_assert(bool_t statement_truth_value, const char* error_msg)
```

Comment:
I think the name "assert" here is misleading.  When reading code, I expect execution to stop when a false value is passed to this function.

Reply:
I don't necessarily agree, but you're free to change that.